src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -55,70 +58,175 @@ static void check(Coordination::Error code, const std::string & path)
 
-void ZooKeeper::init(ZooKeeperArgs args_)
+    const Poco::Net::SocketAddress host_socket_addr{host.host};
+    LOG_TEST(log, "Adding ZooKeeper host {} ({})", host.host, host_socket_addr.toString());
+    return Coordination::ZooKeeper::Node{host_socket_addr, host.original_index, host.secure};
+}
 
+/// TODO get rid of this, converting "host" to "node" is stupid
+Coordination::ZooKeeper::Nodes hostsToNodes(const LoggerPtr & log, std::vector<ShuffleHost> & shuffled_hosts)
 {
-    args = std::move(args_);
-    log = getLogger("ZooKeeper");
+    Coordination::ZooKeeper::Nodes nodes;
 
-    if (args.implementation == "zookeeper")
+    bool dns_error = false;
+    for (auto & host : shuffled_hosts)
     {
-        if (args.hosts.empty())
-            throw KeeperException::fromMessage(Coordination::Error::ZBADARGUMENTS, "No hosts passed to ZooKeeper constructor.");
+        auto & host_string = host.host;
+        try
+        {
+            host.secure = startsWith(host_string, "secure://");
 
-        Coordination::ZooKeeper::Nodes nodes;
-        nodes.reserve(args.hosts.size());
+            if (host.secure)
+                host_string.erase(0, strlen("secure://"));
 
-        /// Shuffle the hosts to distribute the load among ZooKeeper nodes.
-        std::vector<ShuffleHost> shuffled_hosts = shuffleHosts();
-
-        bool dns_error = false;
-        for (auto & host : shuffled_hosts)
+            nodes.emplace_back(hostToNode(log, host));
+        }
+        catch (const Poco::Net::HostNotFoundException & e)
         {
-            auto & host_string = host.host;
-            try
-            {
-                const bool secure = startsWith(host_string, "secure://");
+            /// Most likely it's misconfiguration and wrong hostname was specified
+            LOG_ERROR(log, "Cannot use ZooKeeper host {}, reason: {}", host_string, e.displayText());
+        }
+        catch (const Poco::Net::DNSException & e)
+        {
+            /// Most likely DNS is not available now
+            dns_error = true;
+            LOG_ERROR(log, "Cannot use ZooKeeper host {} due to DNS error: {}", host_string, e.displayText());
+        }
+    }
 
-                if (secure)
-                    host_string.erase(0, strlen("secure://"));
+    if (nodes.empty())
+    {
+        /// For DNS errors we throw exception with ZCONNECTIONLOSS code, so it will be considered as hardware error, not user error
+        if (dns_error)
+            throw KeeperException::fromMessage(
+                Coordination::Error::ZCONNECTIONLOSS, "Cannot resolve any of provided ZooKeeper hosts due to DNS error");
+        else
+            throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot use any of provided ZooKeeper nodes");
+    }
 
-                /// We want to resolve all hosts without DNS cache for keeper connection.
-                Coordination::DNSResolver::instance().removeHostFromCache(host_string);
+    return nodes;
```

Comment:
Also I have concerns about that function. 
It could return not all hosts. As a result we will proceed with the left hosts without even trying recheck deleted hosts.

Imagine if our local az keeper is under recreation. It is delisted from DNS, host has no dns record for some time. We will just proceed without it. Our reconnect logic with `optimal_impl` wont be triggered. When that node become available no one tries to connect to it.

Reply:
I removed this function